src/proxy.js

```diff
@@ -0,0 +1,278 @@
+        checkParamOrThrow(groups, 'opts.groups', 'Maybe [String]');
+        checkParamOrThrow(country, 'opts.country', 'Maybe String');
+        checkParamOrThrow(password, 'opts.password', 'Maybe String');
+        checkParamOrThrow(hostname, 'opts.hostname', 'String', this._getMissingParamErrorMgs('hostname', ENV_VARS.PROXY_HOSTNAME));
+        checkParamOrThrow(port, 'opts.port', 'Number', this._getMissingParamErrorMgs('port', ENV_VARS.PROXY_PORT));
+        this._validateArgumentsStructure(groups, country);
+
+        this.groups = groups;
+        this.country = country;
+        this.password = password;
+        this.hostname = hostname;
+        this.port = port;
+        this.availableGroups = [];
+    }
+
+    /**
+     * Loads proxy password if token is provided and checks access to apify proxy and privided proxy groups.
+     * This function must be called before you can start using the instance in a meaningful way.
+     *
+     * @returns {Promise<void>}
+     */
+    async initialize() {
+        // Set proxy password via API if token is provided
+        await this._setPasswordIfToken();
+
+        // Check if user has access to apify proxy and selected proxy groups
+        await this._checkAccess();
+
+        // TODO: Validate proxyUrl each of custom proxies
+    }
+
+
+    /**
+     * Gets information about proxy and its configuration.
+     * @param {string} sessionId
+     * Apify Proxy session identifier to be used with requests.
+     * All HTTP requests going through the proxy with the same session identifier
+     * will use the same target proxy server (i.e. the same IP address).
+     * The identifier can only contain the following characters: `0-9`, `a-z`, `A-Z`, `"."`, `"_"` and `"~"`.
+     * @return {object} represents information about proxy.
+     */
+    getInfo(sessionId) {
+        const { groups, country, password, port, hostname } = this;
+        const url = this.getUrl(sessionId);
+
+        return {
+            sessionId,
+            url,
+            groups,
+            country,
+            password,
+            hostname,
+            port,
+        };
+    }
+
+    /**
+     * Returns proxy url.
+     * @param {string} sessionId
+     * Apify Proxy session identifier to be used with requests.
+     * All HTTP requests going through the proxy with the same session identifier
+     * will use the same target proxy server (i.e. the same IP address).
+     * The identifier can only contain the following characters: `0-9`, `a-z`, `A-Z`, `"."`, `"_"` and `"~"`.
+     * @return {string} - the proxy url
+     */
+    getUrl(sessionId) {
+        const username = this._getUsername(sessionId);
+        const { password, hostname, port } = this;
+
+        return `${PROTOCOL}://${username}:${password}@${hostname}:${port}`;
+    }
+
+    /**
+     * Returns proxy username.
+     * @return {string} - the proxy username
+     * @param {string} sessionId
+     * @private
+     */
+    _getUsername(sessionId) {
```

Comment:
@mnmkng I do not like it because the username is exactly what it is. The best name would be `getProxyUsername.` However, we are in the `ProxyConfiguration` and the method makes the username of the proxy to use it in its URL and it is a private method (some random developers may think that it is somehow connected to the actual user) but I would keep it that way.  `stringifyOptions` is too general for this. Just my opinion.

Reply:
Ok, sure.